packages/eslint-plugin/src/rules/no-unsafe-type-assertion.ts

```diff
@@ -0,0 +1,156 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import {
+  createRule,
+  getConstrainedTypeAtLocation,
+  getParserServices,
+  isTypeAnyType,
+  isTypeUnknownType,
+  isUnsafeAssignment,
+} from '../util';
+import { isTypeUnchanged } from '../util/isTypeUnchanged';
+
+export default createRule({
+  name: 'no-unsafe-type-assertion',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Disallow type assertions that widen a type',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      unsafeOfAnyTypeAssertion:
+        'Unsafe cast from {{type}} detected: consider using type guards or a safer cast.',
+      unsafeToAnyTypeAssertion:
+        'Unsafe cast to {{type}} detected: consider using a more specific type to ensure safety.',
+      unsafeTypeAssertion:
+        "Unsafe type assertion: type '{{type}}' is more narrow than the original type.",
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+    const compilerOptions = services.program.getCompilerOptions();
+
+    function getAnyTypeName(type: ts.Type): string {
+      return tsutils.isIntrinsicErrorType(type) ? 'error typed' : '`any`';
+    }
+
+    function isObjectLiteralType(type: ts.Type): boolean {
+      return (
+        tsutils.isObjectType(type) &&
+        tsutils.isObjectFlagSet(type, ts.ObjectFlags.ObjectLiteral)
+      );
+    }
+
+    function checkExpression(
+      node: TSESTree.TSAsExpression | TSESTree.TSTypeAssertion,
+    ): void {
+      const expressionType = getConstrainedTypeAtLocation(
+        services,
+        node.expression,
+      );
+      const assertedType = getConstrainedTypeAtLocation(
+        services,
+        node.typeAnnotation,
+      );
+
+      // consider unchanged type as safe
+      const typeIsUnchanged = isTypeUnchanged(
+        compilerOptions,
+        expressionType,
+        assertedType,
+      );
+
+      if (typeIsUnchanged) {
```

Comment:
[Style] Nit: several of these checks both have an explanatory comment _and_ a variable of a similar name. I'd personally remove either or both to be a bit more succinct. I think the code is pretty readable without more explanations.

```suggestion
      if (isTypeUnchanged(compilerOptions, expressionType, assertedType)) {
        return;
      }
```

Just a style nit, not a problem if you ignore it 🙂. The later-on ones around `isUnsafeAssignment` are a little less clear, so I personally might keep one of the two strategies for them in particular...

Reply:
Yeah, I agree that the PR was a bit over-explained with comments. I tried to simplify this and keep the ones I think are important (added one regarding why type-widening is performed which was missing).

Please let me know if this is OK or if I should change anything.